source_code/main_mcu/src/fido2/fido2.c

```diff
@@ -0,0 +1,493 @@
+    uint8_t cred_len_l;
+} __attribute__((packed)) attest_header_t;
+
+typedef struct attested_data_s
+{
+    auth_data_header_t auth_data_header;
+    attest_header_t attest_header;
+    credential_ID_t cred_ID;
+    uint8_t enc_pub_key[ENC_PUB_KEY_LEN];
+    uint32_t enc_PK_len;
+} __attribute__((packed)) attested_data_t;
+
+typedef struct credential_DB_rec_s
+{
+    uint8_t rpID[RPID_LEN];                  //252
+    uint8_t tag[TAG_LEN];                   //16
+    uint8_t user_ID[USER_ID_LEN];            //65
+    uint8_t user_name[USER_NAME_LEN];        //65
+    uint8_t display_name[DISPLAY_NAME_LEN];  //65
+    uint8_t spare;                          //1 (padding)
+    uint8_t priv_key[PRIV_KEY_LEN];         //32
+    uint32_t count;                         //4
+} credential_DB_rec_t;
+
+static bool fido2_room_for_more_creds(uint8_t const *rpID);
+static void fido2_store_credential_in_db(credential_DB_rec_t const *cred_rec);
+static void fido2_update_credential_in_db(credential_DB_rec_t const *cred_rec);
+static credential_DB_rec_t *fido2_get_credential(uint8_t const *rpID, uint8_t const *tag);
+static credential_DB_rec_t *fido2_get_credential_by_index(uint8_t const *rpID, uint32_t index);
+
+/*! \fn     fido2_update_credential_count(uint32_t count)
+*   \brief  Update the credential count. This is to protect against replay attack.
+*           Server will check that count is not going backwards or re-used.
+*   \param  Current count to be updated
+*   \return new count
+*/
+static uint32_t fido2_update_credential_count(uint32_t count)
+{
+    //TBD
+    count++;
+    return count;
+}
+
+/*! \fn     fido2_prompt_user(char const *prompt, bool wait_for_ack, uint32_t timeout_sec)
+*   \brief  Prompt the user and optionally wait for ack from user.
+*   \param  prompt to display
+*   \param  boolean whether to wait for ACK
+*   \param  the amount of time to display message/wait for ack before timeout
+*   \return 0 if accepted or other value if denied/failed
+*/
+static uint32_t fido2_prompt_user(char const *prompt, bool wait_for_ack, uint32_t timeout_sec)
+{
+    uint32_t result = SUCCESS;
+    //TBD
+    //Display request to create cred with name
+    //Wait for ack
+    return result;
+}
+
+/*! \fn     fido2_make_auth_tag(credential_ID_t *cred_ID)
+*   \brief  Make a tag uniquely identifying credential
+*   \param  credential id
+*   \return void
+*/
+static void fido2_make_auth_tag(credential_ID_t *cred_ID)
+{
+    rng_fill_array(cred_ID->tag, TAG_LEN);
+}
+
+/*! \fn     fido2_calc_attestation_signature(uint8_t const *data, int datalen, uint8_t const * client_data_hash, uint8_t * sigbuf)
+*   \brief  Calculate the attestation signature
+*   \param  data to calculate signature over
+*   \param  length of data
+*   \param  hash of client data
+*   \param  out buffer with signature
+*   \return void
+*/
+static void fido2_calc_attestation_signature(uint8_t const *data, int datalen, uint8_t const * client_data_hash, uint8_t * sigbuf)
```

Comment:
could we put a bunch of _static_asserts in order to make sure our buffers are correctly dimensioned?

Reply:
OK, for this one I added _Static_assert(sizeof(hashbuf) >= 32, "Buffer must be 32 bytes or greater");. Will look at other places. too.